lib/ddtrace/buffer.rb

```diff
@@ -8,38 +8,28 @@ class TraceBuffer
     def initialize(max_size)
       @max_size = max_size
 
-      @mutex = Mutex.new()
+      @mutex = Mutex.new
       @traces = []
       @closed = false
     end
 
     # Add a new ``trace`` in the local queue. This method doesn't block the execution
     # even if the buffer is full. In that case, a random trace is discarded.
     def push(trace)
-      @mutex.synchronize do
-        return if @closed
-        len = @traces.length
-        if len < @max_size || @max_size <= 0
-          @traces << trace
-        else
-          # we should replace a random trace with the new one
-          @traces[rand(len)] = trace
-        end
-      end
+      return if @closed
+      len = @traces.length
+      @traces.delete(rand(len)) unless len < @max_size || @max_size <= 0
```

Comment:
Here without mutex, its possible that we'll try to delete from smaller array. But its a case that will not cause any additional loss of information.